### PR TITLE
python310Packages.asyncssh: add optional-dependencies

### PR DESCRIPTION
**Pull request by fabaff**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 212411` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>dvc</li>
    <li>dvc-with-remotes</li>
    <li>python311Packages.scmrepo</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>home-assistant-component-tests.asuswrt</li>
    <li>python310Packages.aioasuswrt</li>
    <li>python310Packages.asyncssh</li>
    <li>python310Packages.scmrepo</li>
    <li>python311Packages.aioasuswrt</li>
    <li>python311Packages.asyncssh</li>
  </ul>
</details>

**Comment by fabaff** [comment]
`gpgme` doesn't support Python 3.11 yet.

**Comment by onny** [comment]
Can you tell me how does this optional dependency thing work in Python? The build process doesn't require them to be present? Is it build with all optional dependencies? How can I disable them?

**Comment by fabaff** [comment]
> Can you tell me how does this optional dependency thing work in Python?

The "base module" (sorry, I don't recall the exact terminology) only contains the core functionality. If you would like to use `asyncssh` then you would do:

```bash
$ python3 -m pip install asyncssh
```

For FIDO support one would need to mention that you want support for those features.

```bash
python3 -m pip install asyncssh[fido2]
```

This will get you `asyncssh` plus everything that is needed to have FIDO2 support. 

More details: https://peps.python.org/pep-0508/

> The build process doesn't require them to be present?

No, as they are optional (listed in e.g., `extras_require`) aka not part of `install_require`. But they could be needed for the test suite (e.g., `++ passthru.optional-dependencies.fido2`).

> Is it build with all optional dependencies?

`asyncssh` included all dependencies so far in `propagatedBuildInputs`. This means that the output includes all dependencies and had support for all features. With `optional-dependencies` present, other modules can now only pull-in the parts they need. 

> How can I disable them?

For the module itself you would simply not mention the dependencies. This is usually done if the dependency is not present.

File: pkgs/development/python-modules/asyncssh/default.nix [hunk 1]
```diff
@@ -31,23 +31,43 @@ buildPythonPackage rec {
...
     openssh
     openssl
     pytestCheckHook
-  ];
+  ] ++ passthru.optional-dependencies.bcrypt
+  ++ passthru.optional-dependencies.fido2
+  ++ passthru.optional-dependencies.gssapi
+  ++ passthru.optional-dependencies.libnacl
+  ++ passthru.optional-dependencies.pkcs11
+  ++ passthru.optional-dependencies.pyOpenSSL;
```
Comment (mweinelt) [inline pkgs/development/python-modules/asyncssh/default.nix:70]:
```suggestion
  ] ++ lib.flatten (builtins.attrValues passthru.optional-dependencies);
```

**Comment by onny** [comment]
@fabaff Thank you for the kind explanation, good to know this about python packaging!

**Comment by fabaff** [comment]
Still waiting on `gpgme`.

File: pkgs/development/python-modules/asyncssh/default.nix (hunk 1 above)
Comment (fabaff) [reply pkgs/development/python-modules/asyncssh/default.nix:70]: Thanks, updated.

**Comment by fabaff** [comment]
Result of `nixpkgs-review pr 212411` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>dvc</li>
    <li>dvc-with-remotes</li>
    <li>python311Packages.scmrepo</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>home-assistant-component-tests.asuswrt</li>
    <li>python310Packages.aioasuswrt</li>
    <li>python310Packages.asyncssh</li>
    <li>python310Packages.scmrepo</li>
    <li>python311Packages.aioasuswrt</li>
    <li>python311Packages.asyncssh</li>
  </ul>
</details>

**Pull request merged**